In [24]:
from IPython.core.interactiveshell import InteractiveShell
InteractiveShell.ast_node_interactivity = "all" 

# 4.1 Classifying with Bayesian decision theory

* 이전까지는 정보를 가지고 분류 하다가가 확률을 가지고 분류함 

- 장점 : 소량의 데이터, 여러 개의 분류 항목을 다룸
- 단점 : 입력데이터에 따라 민감함
- 적용 : 명목형 값


* 어디에 속할 확률이 높은가 여부로 판별함
    - if p1(x, y) > p2(x, y), then the class is 1.
    - if p2(x, y) > p1(x, y), then the class is 2.


* Byes?
    * 베이지안확률 이라고하는 분류 사용하기 편하고 인기가 있음 
    * 확실하지 않은 상태의 사전지식과 논리를 적용하는 것이 가능 
    * 빈발 확률 
    * 데이터로부터 결론을 도출해 내고 논리와 사전 지식을 허용하지 않음
    
    
# 4.2 Conditional probability
조건부확률

P(A|B) = p(A∩B)/p(B)<br>
p(A∩B) = p(B|A)*p(A)

    -> P(A|B) = p(B|A)*p(A) / p(B)
    
    
# 4.3 Classifying with conditional probabilities

If P(c1|x, y) > P(c2|x, y), the class is c1. <br>
If P(c1|x, y) < P(c2|x, y), the class is c2.

# 4.4 Document classification with navie Bayes
General approach to naïve Bayes
1. Collect: 
    - Any method. We’ll use RSS feeds in this chapter.
2. Prepare: 
    - Numeric or Boolean values are needed.
3. Analyze: 
    - With many features, plotting features isn’t helpful. Looking at histograms is a better idea.
4. Train: 
    - Calculate the conditional probabilities of the independent features.
5. Test: 
    - Calculate the error rate.
6. Use: 
    - One common application of naïve Bayes is document classification. 
    - You can use naïve Bayes in any classification setting. It doesn’t have to be text.


- 만약에 하나의 속성에 대해 N개의 표본이 필요하다면, 10개의 속성을 위해서는 N^10개의 표본이 필요하게 되며, 어휘 목록에 있는 1000개의 속성을 위해서는 N^1000개의 표본이 필요하게 된다. 숫자는 매우 빠르게 증가한다.
- 속성들이 서로 독립적이라고 가정하면, N^1000개의 데이터는 1000 x N개로 줄어들게 된다. 독립이라는 것은 통계적인 독립을 의미한다.
- 1000개의 단어를 확인할 필요 없이 10개 또는 20개의 단어만을 확인해 보아도 될 것이다. 이러한 가정에 약간의 결함이 있는데도 불구하고, 나이브 베이스는 일을 잘 수행한다.

# 4.5 Classifying text with Python
텍스트 문서에서 속성을 추출하기 위해서는 텍스트를 분할해야 한다. 

- 속성은 텍스트 문서에서 얻을 수 있는 토큰(token)들이 될 것이다.
- 텍스트의 모든 조각을 토큰 벡터로 줄일 것이다. 문서에서 1은 현존하는 토큰을 0은 존재하지 않는 토큰을 나타낸다.

## 4.5.1 Prepare: making word vectors from text
### Listing 4.1 Word list to vector function

In [25]:
def loadDataSet():
    # 소문자로 만들고 토큰생성!! 
    postingList=[['my', 'dog', 'has', 'flea', \
                'problems', 'help', 'please'],
                ['maybe', 'not', 'take', 'him', \
                'to', 'dog', 'park', 'stupid'],
                ['my', 'dalmation', 'is', 'so', 'cute', \
                'I', 'love', 'him'],
                ['stop', 'posting', 'stupid', 'worthless', 'garbage'],
                ['mr', 'licks', 'ate', 'my', 'steak', 'how',\
                'to', 'stop', 'him'],
                ['quit', 'buying', 'worthless', 'dog', 'food', 'stupid']]
    classVec = [0,1,0,1,0,1] #1 is abusive, 0 not
    
    return postingList,classVec

In [26]:
def createVocabList(dataSet):
    vocabSet = set([])
    for document in dataSet:
        vocabSet = vocabSet | set(document) # set() : remove duplicated word  '|'는 없는 것만 합해준다!!!
        # ex) ('dog','flea','has','help','my','please','problem') sort도 된다. 이걸 반복해 추가해 간다!!  
    return list(vocabSet)

In [27]:
def setOfWords2Vec(vocabList, inputSet): # 단어의 벡터화!! 
    returnVec = [0]*len(vocabList)
    for word in inputSet:
        if word in vocabList:
            returnVec[vocabList.index(word)] = 1
        else: print "the word: %s is not in my Vocabulary!" % word
    return returnVec

- 위 bayes.py 안에 있는 코드다!!

In [28]:
import bayes

In [29]:
reload(bayes)

<module 'bayes' from 'bayes.pyc'>

In [30]:
list_of_post, class_label = bayes.loadDataSet()
list_of_post
class_label

[['my', 'dog', 'has', 'flea', 'problems', 'help', 'please'],
 ['maybe', 'not', 'take', 'him', 'to', 'dog', 'park', 'stupid'],
 ['my', 'dalmation', 'is', 'so', 'cute', 'I', 'love', 'him'],
 ['stop', 'posting', 'stupid', 'worthless', 'garbage'],
 ['mr', 'licks', 'ate', 'my', 'steak', 'how', 'to', 'stop', 'him'],
 ['quit', 'buying', 'worthless', 'dog', 'food', 'stupid']]

[0, 1, 0, 1, 0, 1]

In [31]:
my_vocab_list = bayes.createVocabList_lec(list_of_post)

document: ['my', 'dog', 'has', 'flea', 'problems', 'help', 'please']
set(document): set(['help', 'please', 'problems', 'dog', 'flea', 'has', 'my'])
vocabSet: set(['help', 'my', 'flea', 'problems', 'has', 'please', 'dog'])

document: ['maybe', 'not', 'take', 'him', 'to', 'dog', 'park', 'stupid']
set(document): set(['maybe', 'park', 'dog', 'to', 'stupid', 'take', 'not', 'him'])
vocabSet: set(['not', 'help', 'park', 'problems', 'please', 'dog', 'to', 'stupid', 'take', 'maybe', 'flea', 'has', 'my', 'him'])

document: ['my', 'dalmation', 'is', 'so', 'cute', 'I', 'love', 'him']
set(document): set(['cute', 'love', 'I', 'is', 'dalmation', 'so', 'my', 'him'])
vocabSet: set(['cute', 'love', 'help', 'I', 'park', 'is', 'problems', 'not', 'dalmation', 'flea', 'him', 'maybe', 'please', 'dog', 'to', 'stupid', 'so', 'take', 'has', 'my'])

document: ['stop', 'posting', 'stupid', 'worthless', 'garbage']
set(document): set(['worthless', 'stupid', 'stop', 'posting', 'garbage'])
vocabSet: set(['cute', 'lov

- docuemnt를 set 하고 vocabSet에 '|'로 추가해 나간다!! 

In [32]:
len(my_vocab_list)

32

In [33]:
reload(bayes)

<module 'bayes' from 'bayes.pyc'>

In [34]:
list_of_post[0]
bayes.setOfWords2Vec_lec(my_vocab_list, list_of_post[0])

['my', 'dog', 'has', 'flea', 'problems', 'help', 'please']

returnVec: [0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0]
->vocabList: ['cute', 'love', 'help', 'garbage', 'quit', 'I', 'problems', 'is', 'park', 'stop', 'flea', 'dalmation', 'licks', 'food', 'not', 'him', 'buying', 'posting', 'has', 'worthless', 'ate', 'to', 'maybe', 'please', 'dog', 'how', 'stupid', 'so', 'take', 'mr', 'steak', 'my']
word: my

---->word: my
---->vocabList.index(word): 31
---->returnVec[vocabList.index(word)]: 0
---->returnVec[vocabList.insdex(word)]: 1

word: dog

---->word: dog
---->vocabList.index(word): 24
---->returnVec[vocabList.index(word)]: 0
---->returnVec[vocabList.insdex(word)]: 1

word: has

---->word: has
---->vocabList.index(word): 18
---->returnVec[vocabList.index(word)]: 0
---->returnVec[vocabList.insdex(word)]: 1

word: flea

---->word: flea
---->vocabList.index(word): 10
---->returnVec[vocabList.index(word)]: 0
---->returnVec[vocabList.insdex(word)]: 1

word: problems

---->word: problems
---->vocabLi

[0,
 0,
 1,
 0,
 0,
 0,
 1,
 0,
 0,
 0,
 1,
 0,
 0,
 0,
 0,
 0,
 0,
 0,
 1,
 0,
 0,
 0,
 0,
 1,
 1,
 0,
 0,
 0,
 0,
 0,
 0,
 1]

word: my

----> word: my

----> vocabList.index(word): 31

----> returnVec[vocabList.index(word)]: 0

----> returnVec[vocabList.insdex(word)]: 1

- my가 vocabList에 있는데 31번째에 있다!!

##### 이렇게 word_Vector 생성함!! 

In [35]:
list_of_post[1]

['maybe', 'not', 'take', 'him', 'to', 'dog', 'park', 'stupid']

In [36]:
bayes.setOfWords2Vec(my_vocab_list, list_of_post[1])

[0,
 0,
 0,
 0,
 0,
 0,
 0,
 0,
 1,
 0,
 0,
 0,
 0,
 0,
 1,
 1,
 0,
 0,
 0,
 0,
 0,
 1,
 1,
 0,
 1,
 0,
 1,
 0,
 1,
 0,
 0,
 0]

## 4.5.2 Train: calculating probabilities from word vectors
1. 이 단어가 나타날 확률
2. 폭력적인 문서일 확률
3. 폭력적인 문서에서 그 단어가 나올 확률


- 위 3가지를 가지고 이 단어가 나타날 때 이 문서가 폭력적인 문서일 확률을 구한다!! 
- 독립이라는 가정 덕분에, 폭력적인 문서일 때 특정 단어들이 나올 확률을 구할 수 있다 곱으로!!

↓ psuedo code

### Listing 4.2 Naïve Bayes classifier training function

In [37]:
def trainNB0(trainMatrix,trainCategory):
    numTrainDocs = len(trainMatrix)
    numWords = len(trainMatrix[0])
    pAbusive = sum(trainCategory)/float(numTrainDocs)
    # 확률 초기화
    p0Num = zeros(numWords); p1Num = zeros(numWords)
    p0Denom = 0.0; p1Denom = 0.0
    
    for i in range(numTrainDocs):
        if trainCategory[i] == 1:
            # 벡터 추가
            p1Num += trainMatrix[i]
            p1Denom += sum(trainMatrix[i])
        else:
            p0Num += trainMatrix[i]
            p0Denom += sum(trainMatrix[i])
    # 원소 나누기
    p1Vect = p1Num/p1Denom #change to log() # P(Wi/C1)
    p0Vect = p0Num/p0Denom #change to log() # P(Wi/C0)
    return p0Vect,p1Vect,pAbusive

In [38]:
reload(bayes)

<module 'bayes' from 'bayes.pyc'>

In [39]:
list_of_post, class_label = bayes.loadDataSet()

In [41]:
my_vocab_list = bayes.createVocabList(list_of_post)

In [51]:
trainMat = []
for post in list_of_post:
    trainMat.append(bayes.setOfWords2Vec(my_vocab_list, post))
# listComprehension 도 가능하지!!

In [52]:
trainMat2 = [bayes.setOfWords2Vec(my_vocab_list, post) for post in list_of_post]

In [53]:
trainMat == trainMat2

True

In [67]:
print trainMat[0]
print trainMat[1]

[0, 0, 1, 0, 0, 0, 1, 0, 0, 0, 1, 0, 0, 0, 0, 0, 0, 0, 1, 0, 0, 0, 0, 1, 1, 0, 0, 0, 0, 0, 0, 1]
[0, 0, 0, 0, 0, 0, 0, 0, 1, 0, 0, 0, 0, 0, 1, 1, 0, 0, 0, 0, 0, 1, 1, 0, 1, 0, 1, 0, 1, 0, 0, 0]


In [80]:
reload(bayes)

<module 'bayes' from 'bayes.py'>

In [81]:
print my_vocab_list

['cute', 'love', 'help', 'garbage', 'quit', 'I', 'problems', 'is', 'park', 'stop', 'flea', 'dalmation', 'licks', 'food', 'not', 'him', 'buying', 'posting', 'has', 'worthless', 'ate', 'to', 'maybe', 'please', 'dog', 'how', 'stupid', 'so', 'take', 'mr', 'steak', 'my']


In [82]:
print trainMat
class_label

[[0, 0, 1, 0, 0, 0, 1, 0, 0, 0, 1, 0, 0, 0, 0, 0, 0, 0, 1, 0, 0, 0, 0, 1, 1, 0, 0, 0, 0, 0, 0, 1], [0, 0, 0, 0, 0, 0, 0, 0, 1, 0, 0, 0, 0, 0, 1, 1, 0, 0, 0, 0, 0, 1, 1, 0, 1, 0, 1, 0, 1, 0, 0, 0], [1, 1, 0, 0, 0, 1, 0, 1, 0, 0, 0, 1, 0, 0, 0, 1, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 1, 0, 0, 0, 1], [0, 0, 0, 1, 0, 0, 0, 0, 0, 1, 0, 0, 0, 0, 0, 0, 0, 1, 0, 1, 0, 0, 0, 0, 0, 0, 1, 0, 0, 0, 0, 0], [0, 0, 0, 0, 0, 0, 0, 0, 0, 1, 0, 0, 1, 0, 0, 1, 0, 0, 0, 0, 1, 1, 0, 0, 0, 1, 0, 0, 0, 1, 1, 1], [0, 0, 0, 0, 1, 0, 0, 0, 0, 0, 0, 0, 0, 1, 0, 0, 1, 0, 0, 1, 0, 0, 0, 0, 1, 0, 1, 0, 0, 0, 0, 0]]


[0, 1, 0, 1, 0, 1]

In [83]:
bayes.trainNB0_lec(trainMat, class_label)

prob. of abusive: 0.5

-->i: 0
-->if trainCategory[0] == 1
--> trainCategory: [0, 1, 0, 1, 0, 1]
----> trainMatrix[0]: [0, 0, 1, 0, 0, 0, 1, 0, 0, 0, 1, 0, 0, 0, 0, 0, 0, 0, 1, 0, 0, 0, 0, 1, 1, 0, 0, 0, 0, 0, 0, 1]
----> p0Num: [ 0.  0.  1.  0.  0.  0.  1.  0.  0.  0.  1.  0.  0.  0.  0.  0.  0.  0.
  1.  0.  0.  0.  0.  1.  1.  0.  0.  0.  0.  0.  0.  1.]
----> p0Denom: 7.0

-->i: 1
-->if trainCategory[1] == 1
--> trainCategory: [0, 1, 0, 1, 0, 1]
----> trainMatrix[1]: [0, 0, 0, 0, 0, 0, 0, 0, 1, 0, 0, 0, 0, 0, 1, 1, 0, 0, 0, 0, 0, 1, 1, 0, 1, 0, 1, 0, 1, 0, 0, 0]
----> p1Num: [ 0.  0.  0.  0.  0.  0.  0.  0.  1.  0.  0.  0.  0.  0.  1.  1.  0.  0.
  0.  0.  0.  1.  1.  0.  1.  0.  1.  0.  1.  0.  0.  0.]
----> p1Denom: 8.0

-->i: 2
-->if trainCategory[2] == 1
--> trainCategory: [0, 1, 0, 1, 0, 1]
----> trainMatrix[2]: [1, 1, 0, 0, 0, 1, 0, 1, 0, 0, 0, 1, 0, 0, 0, 1, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 1, 0, 0, 0, 1]
----> p0Num: [ 1.  1.  1.  0.  0.  1.  1.  1.  0.  0.  1.  1.  0.  0. 

(array([ 0.04166667,  0.04166667,  0.04166667,  0.        ,  0.        ,
         0.04166667,  0.04166667,  0.04166667,  0.        ,  0.04166667,
         0.04166667,  0.04166667,  0.04166667,  0.        ,  0.        ,
         0.08333333,  0.        ,  0.        ,  0.04166667,  0.        ,
         0.04166667,  0.04166667,  0.        ,  0.04166667,  0.04166667,
         0.04166667,  0.        ,  0.04166667,  0.        ,  0.04166667,
         0.04166667,  0.125     ]),
 array([ 0.        ,  0.        ,  0.        ,  0.05263158,  0.05263158,
         0.        ,  0.        ,  0.        ,  0.05263158,  0.05263158,
         0.        ,  0.        ,  0.        ,  0.05263158,  0.05263158,
         0.05263158,  0.05263158,  0.05263158,  0.        ,  0.10526316,
         0.        ,  0.05263158,  0.05263158,  0.        ,  0.10526316,
         0.        ,  0.15789474,  0.        ,  0.05263158,  0.        ,
         0.        ,  0.        ]),
 0.5)

## 4.5.3 Test: modifying the classifier for real-world conditions

실제 조건을 반영하기 위해서 
- p1Vect, p0Vect 에서 확률이 0인 것들을 수정해 줘야한다!! 
    - 그래서 모든 단어가 한번은 나온다고 처리해 준다!! 
        - p0Num = ones(numWords) 
        - p1Num = ones(numWords)
            - ↑zeros가 아닌 ones를 사용한다
        - p0Denom = 2.0; p1Denom = 2.0
            - ↑ 초기값 0이 아닌 2로하는 이유는, Abusive 냐 아니냐 2가지 이기 때문에 2로 준것이다!! / 패널티 개념이다!!


- 문장이 너무 많으면, 단어의 빈도가 작으면 그 단어가 나올 확률이 매우 작아진다. 매우 작은 값끼리 곱하면 언더플로우(underflow) 문제가 발생하게 된다!!!
    - 효과
        - 스케일을 줄여준다.
        - log 지수를 쓰면 곱셉이 덧셈으로 변하기 때문에 계산이 간편해짐!!

In [84]:
reload(bayes)

<module 'bayes' from 'bayes.py'>

In [85]:
bayes.trainNB0_lec(trainMat, class_label)

prob. of abusive: 0.5

-->i: 0
-->if trainCategory[0] == 1
--> trainCategory: [0, 1, 0, 1, 0, 1]
----> trainMatrix[0]: [0, 0, 1, 0, 0, 0, 1, 0, 0, 0, 1, 0, 0, 0, 0, 0, 0, 0, 1, 0, 0, 0, 0, 1, 1, 0, 0, 0, 0, 0, 0, 1]
----> p0Num: [ 1.  1.  2.  1.  1.  1.  2.  1.  1.  1.  2.  1.  1.  1.  1.  1.  1.  1.
  2.  1.  1.  1.  1.  2.  2.  1.  1.  1.  1.  1.  1.  2.]
----> p0Denom: 9.0

-->i: 1
-->if trainCategory[1] == 1
--> trainCategory: [0, 1, 0, 1, 0, 1]
----> trainMatrix[1]: [0, 0, 0, 0, 0, 0, 0, 0, 1, 0, 0, 0, 0, 0, 1, 1, 0, 0, 0, 0, 0, 1, 1, 0, 1, 0, 1, 0, 1, 0, 0, 0]
----> p1Num: [ 1.  1.  1.  1.  1.  1.  1.  1.  2.  1.  1.  1.  1.  1.  2.  2.  1.  1.
  1.  1.  1.  2.  2.  1.  2.  1.  2.  1.  2.  1.  1.  1.]
----> p1Denom: 10.0

-->i: 2
-->if trainCategory[2] == 1
--> trainCategory: [0, 1, 0, 1, 0, 1]
----> trainMatrix[2]: [1, 1, 0, 0, 0, 1, 0, 1, 0, 0, 0, 1, 0, 0, 0, 1, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 1, 0, 0, 0, 1]
----> p0Num: [ 2.  2.  2.  1.  1.  2.  2.  2.  1.  1.  2.  2.  1.  1.

(array([-2.56494936, -2.56494936, -2.56494936, -3.25809654, -3.25809654,
        -2.56494936, -2.56494936, -2.56494936, -3.25809654, -2.56494936,
        -2.56494936, -2.56494936, -2.56494936, -3.25809654, -3.25809654,
        -2.15948425, -3.25809654, -3.25809654, -2.56494936, -3.25809654,
        -2.56494936, -2.56494936, -3.25809654, -2.56494936, -2.56494936,
        -2.56494936, -3.25809654, -2.56494936, -3.25809654, -2.56494936,
        -2.56494936, -1.87180218]),
 array([-3.04452244, -3.04452244, -3.04452244, -2.35137526, -2.35137526,
        -3.04452244, -3.04452244, -3.04452244, -2.35137526, -2.35137526,
        -3.04452244, -3.04452244, -3.04452244, -2.35137526, -2.35137526,
        -2.35137526, -2.35137526, -2.35137526, -3.04452244, -1.94591015,
        -3.04452244, -2.35137526, -2.35137526, -3.04452244, -1.94591015,
        -3.04452244, -1.65822808, -3.04452244, -2.35137526, -3.04452244,
        -3.04452244, -3.04452244]),
 0.5)

- 검증 ()

In [86]:

def classifyNB_lec(vec2Classify, p0Vec, p1Vec, pClass1): 
    p1 = sum(vec2Classify * p1Vec) + log(pClass1)    #element-wise mult
    p0 = sum(vec2Classify * p0Vec) + log(1.0 - pClass1)
    if p1 > p0:
        return 1
    else: 
        return 0

- vec2Classify, P(c1|w[t]) 에서 w[t]를 나타낸다. 테스트할 word들의 vector!!
    - 이 테스트할 word 들이 테스트 하고 싶은 글에 들어 있는 거지!! 
    - 그래서 그 글을 테스트 하는 거다!! 


- pClass1, P(c1) 이다! - 폭력적인 글일 확률


- P(c1|w[t]) = (P(w|c1) * P(c1)) / P(w)      
    - P(w) = P(w1,w2 ... ,w31)


- sum(vec2Classify * p1Vec) + log(pClass1)
    - 덧셈을 하는 이유는 log를 취해서 p1Vec을 구했었기 때문에!!     


- Q. 그러면 32개의 정해진 단어에 대해서만 확률을 가지고 있기 때문에 테스트 하는 문장들도 그 32개의 단어만 count 해서 TEST 하겠군! ㅇㅇ

In [87]:
reload(bayes)

<module 'bayes' from 'bayes.py'>

In [88]:
p0v, p1v, pAbusive = bayes.trainNB0(trainMat, class_label)

In [89]:
testEntry = ['love','my','dalmation']

In [94]:
testVec = bayes.setOfWords2Vec(my_vocab_list, testEntry)

In [95]:
print testVec

[0, 1, 0, 0, 0, 0, 0, 0, 0, 0, 0, 1, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 1]


In [96]:
bayes.classifyNB_lec(testVec, p0v, p1v, pAbusive)

vec2Classify: [0, 1, 0, 0, 0, 0, 0, 0, 0, 0, 0, 1, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 1]
p1Vec: [-3.04452244 -3.04452244 -3.04452244 -2.35137526 -2.35137526 -3.04452244
 -3.04452244 -3.04452244 -2.35137526 -2.35137526 -3.04452244 -3.04452244
 -3.04452244 -2.35137526 -2.35137526 -2.35137526 -2.35137526 -2.35137526
 -3.04452244 -1.94591015 -3.04452244 -2.35137526 -2.35137526 -3.04452244
 -1.94591015 -3.04452244 -1.65822808 -3.04452244 -2.35137526 -3.04452244
 -3.04452244 -3.04452244]
p0Vec: [-2.56494936 -2.56494936 -2.56494936 -3.25809654 -3.25809654 -2.56494936
 -2.56494936 -2.56494936 -3.25809654 -2.56494936 -2.56494936 -2.56494936
 -2.56494936 -3.25809654 -3.25809654 -2.15948425 -3.25809654 -3.25809654
 -2.56494936 -3.25809654 -2.56494936 -2.56494936 -3.25809654 -2.56494936
 -2.56494936 -2.56494936 -3.25809654 -2.56494936 -3.25809654 -2.56494936
 -2.56494936 -1.87180218]
pClass1: 0.5
p1: -9.82671449373, p0: -7.69484807238


0

##### p1: -9.82671449373, p0: -7.69484807238
p0 > p1 이기 때문에 이 글은 폭력적이라고 할 수 없다!!! 

- classifyNB_lec 이 포함된 코드가 testingNB() 이다.

In [97]:
bayes.testingNB()

['love', 'my', 'dalmation'] classified as:  0
['stupid', 'garbage'] classified as:  1


↑ 첫번째는 폭력적X / 두번째는 폭력적O 글이다!!! 

In [98]:
testEntry = ['I','dumped','garbage','for','my','cute','dog']

In [101]:
testVec = bayes.setOfWords2Vec(my_vocab_list, testEntry)
print testVec
# 없는 단어도 예외처리 하였음!!

the word: dumped is not in my Vocabulary!
the word: for is not in my Vocabulary!
[1, 0, 0, 1, 0, 1, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 1, 0, 0, 0, 0, 0, 0, 1]


In [102]:
bayes.classifyNB(testVec, p0v, p1v, pAbusive)
# 폭력적인 글이 아니라고 판단했다!! 

0

## 4.5.4 Prepare: the bag-of-words document model

In [103]:
def bagOfWords2VecMN(vocabList, inputSet):
    returnVec = [0]*len(vocabList)
    for word in inputSet:
        if word in vocabList:
            returnVec[vocabList.index(word)] += 1
        else: 
            print "the word: %s is not in my Vocabulary!" % word
    return returnVec

# 4.6 Example: classifying spam email with naïve Bayes



Example: using naïve Bayes to classify email
1. Collect: 
    - Text files provided.
2. Prepare: 
    - Parse text into token vectors.
3. Analyze: 
    - Inspect the tokens to make sure parsing was done correctly.
4. Train: 
    - Use trainNB0() that we created earlier.
5. Test: 
    - Use classifyNB() and create a new testing function to calculate the error rate over a set of documents.
6. Use: 
    - Build a complete program that will classify a group of documents and print misclassified documents to the screen.

## 4.6.1 Prepare: tokenizing text

In [105]:
mySent='This book is the best book on Python or M.L. I have ever laid eyes upon.'
print mySent.split()


['This', 'book', 'is', 'the', 'best', 'book', 'on', 'Python', 'or', 'M.L.', 'I', 'have', 'ever', 'laid', 'eyes', 'upon.']


In [107]:
import re
regEx = re.compile('\\W*') # 이 패턴은, python2에서는 \\해야 \로 인식한다 '' 안에서는!! 
# W는 White space를 지칭한다!! . / \t / , / \n / \r 등 을 뜻한다!! 
listOfTokens = regEx.split(mySent)
print listOfTokens

['This', 'book', 'is', 'the', 'best', 'book', 'on', 'Python', 'or', 'M', 'L', 'I', 'have', 'ever', 'laid', 'eyes', 'upon', '']


- ↑ 마침표 제거를 했네.. 단 M.L을 분리한 것은 아쉽다.. 수작업 필요!!! 
- ↓ list comprehension!! 

In [109]:
x = [tok for tok in listOfTokens if len(tok) > 0]
print x

['This', 'book', 'is', 'the', 'best', 'book', 'on', 'Python', 'or', 'M', 'L', 'I', 'have', 'ever', 'laid', 'eyes', 'upon']


In [110]:
print [tok.lower() for tok in listOfTokens if len(tok) > 0]

['this', 'book', 'is', 'the', 'best', 'book', 'on', 'python', 'or', 'm', 'l', 'i', 'have', 'ever', 'laid', 'eyes', 'upon']


In [111]:
emailText = open('email/ham/6.txt').read()
listOfTokens=regEx.split(emailText)

## 4.6.2 Test: cross validation with naïve Bayes

### Listing 4.5 File parsing and full spam test functions

In [112]:
def textParse(bigString):
    import re
    listOfTokens = re.split(r'\W*', bigString) # r을 이용해서 이후 특수문자 없이 그 자체로 인식하라는 뜻!! 
    return [tok.lower() for tok in listOfTokens if len(tok) > 2] # 3글자 이상은 되어야 의미 있을 것이라고 인식!!!

In [149]:
def spamTest():
    docList=[]; classList = []; fullText =[] # 초기화! 
    for i in range(1,26): # 1 ~ 25 번 파일!! 
        print "#{}\n".format(i)
        wordList = textParse(open('email/spam/%d.txt' % i).read()) # digits
        docList.append(wordList)
        fullText.extend(wordList)
        # append VS extend : 아래 정리함!! 
        classList.append(1) # 스팸메일이다 뜻임
        print "* for spam email"
        print " wordList: {}".format(wordList)
        print " docList: {} ".format(docList)
        print " classList: {}".format(classList)
        print " funllText: {}".format(fullText)
        
        wordList = textParse(open('email/ham/%d.txt' % i).read())
        docList.append(wordList)
        fullText.extend(wordList)
        classList.append(0) # 스팸메일 아니다고 지정! 
        # 1,0,1,0,1,0 이런식으로 50개 저장됨!! 
        print "* for ham email"
        print " wordList: {}".format(wordList)
        print " docList: {} ".format(docList)
        print " classList: {}".format(classList)
        print " funllText: {}".format(fullText)
    vocabList = createVocabList(docList) # 중첩 list를 받아서 set 적용함! 등등..
    trainingSet = range(50); testSet=[]
    # 0~49의 Index 값 중에서 random으로 10개 제거함!! 
    # 그 10개 Index는 testSet으로 활용!! 
    for i in range(10):
        randIndex = int(random.uniform(0,len(trainingSet))) # trainingSet 개수가 변하기 때문에 중복값을 제거 할 일이 없다!! 
        testSet.append(trainingSet[randIndex]) # 제거하기 전에 testSet으로 쓸 Index 지정함!!! 
        del(trainingSet[randIndex])

    trainMat=[]; trainClasses = []
    for docIndex in trainingSet: # 40번! 
        trainMat.append(bagOfWords2Vec(vocabList, docList[docIndex]))
        # bagOfWords 이기 때문에 중복word counting 증가한다!! 
        trainClasses.append(classList[docIndex])
        
    p0V,p1V,pSpam = trainNB0(array(trainMat),array(trainClasses))
    # ↑ p0V,p1V,pSpam 구함, 이전 정리 참고!! 
    print "p0V: {}\np1V: {}\npSpam: {}".format(p0v, p1v, pSpam)
    
    errorCount = 0
    for docIndex in testSet:
        wordVector = bagOfWords2Vec(vocabList, docList[docIndex])

        # classifyNB 계산후 실제값과 비교!! 
        if classifyNB(array(wordVector),p0V,p1V,pSpam) != classList[docIndex]:
            errorCount += 1
            print "classification error", doList[doIndex]

            
    print 'the error rate is: ',float(errorCount)/len(testSet)

- append VS extend

In [128]:
t1 = ['a','bc','def']; t2=[1,2,'hello'] 

In [139]:
t = [];

In [140]:
t.append(t1)

In [141]:
print t

[['a', 'bc', 'def']]


In [142]:
t.append(t2); t

[['a', 'bc', 'def'], [1, 2, 'hello']]

list 자체로 input 됨. 중첩 리스트가 되지!! 

In [143]:
tt = []

In [144]:
tt.extend(t1); tt

['a', 'bc', 'def']

In [145]:
tt.extend(t2); tt

['a', 'bc', 'def', 1, 2, 'hello']

list 가 풀려서 값만 input 된다!! 

In [147]:
reload(bayes)

<module 'bayes' from 'bayes.py'>

In [148]:
bayes.spamTest()

#1

* for spam email
 wordList: ['codeine', '15mg', 'for', '203', 'visa', 'only', 'codeine', 'methylmorphine', 'narcotic', 'opioid', 'pain', 'reliever', 'have', '15mg', '30mg', 'pills', '15mg', 'for', '203', '15mg', 'for', '385', '15mg', 'for', '562', 'visa', 'only']
 docList: [['codeine', '15mg', 'for', '203', 'visa', 'only', 'codeine', 'methylmorphine', 'narcotic', 'opioid', 'pain', 'reliever', 'have', '15mg', '30mg', 'pills', '15mg', 'for', '203', '15mg', 'for', '385', '15mg', 'for', '562', 'visa', 'only']] 
 classList: [1]
 funllText: ['codeine', '15mg', 'for', '203', 'visa', 'only', 'codeine', 'methylmorphine', 'narcotic', 'opioid', 'pain', 'reliever', 'have', '15mg', '30mg', 'pills', '15mg', 'for', '203', '15mg', 'for', '385', '15mg', 'for', '562', 'visa', 'only']
* for ham email
 wordList: ['peter', 'with', 'jose', 'out', 'town', 'you', 'want', 'meet', 'once', 'while', 'keep', 'things', 'going', 'and', 'some', 'interesting', 'stuff', 'let', 'know', 'eugene']
 docList: [['codeine

In [150]:
import numpy as np

In [153]:
[int(np.random.uniform(0,50)) for _ in range(10)] 
# _는 받아도 쓰지 않겠다. 그냥 반복 횟수로만 사용 하겠다.
randIndex = int(np.random.uniform(0,50))
randIndex

[21, 4, 48, 29, 48, 49, 36, 9, 13, 13]

32

- print "p0V: {}\np1V: {}\npSpam: {}".format(p0v, p1v, pSpam)

확인 ↓

In [156]:
reload(bayes)

<module 'bayes' from 'bayes.py'>

In [157]:
bayes.spamTest()

p0V: [-5.42274494 -5.19960139 -6.11589213 -6.11589213 -6.80903931 -6.80903931
 -6.11589213 -6.11589213 -6.11589213 -6.80903931 -6.11589213 -6.80903931
 -6.11589213 -6.11589213 -6.80903931 -6.11589213 -6.11589213 -6.80903931
 -5.71042702 -6.11589213 -6.80903931 -6.80903931 -6.11589213 -6.80903931
 -6.11589213 -6.11589213 -6.11589213 -6.80903931 -6.11589213 -5.71042702
 -5.71042702 -5.71042702 -6.11589213 -6.11589213 -6.11589213 -6.80903931
 -5.19960139 -6.80903931 -6.11589213 -6.11589213 -5.71042702 -6.80903931
 -6.11589213 -6.80903931 -6.11589213 -6.11589213 -6.80903931 -6.11589213
 -5.71042702 -6.11589213 -6.11589213 -5.71042702 -6.11589213 -6.11589213
 -6.11589213 -6.80903931 -6.80903931 -5.42274494 -5.42274494 -6.80903931
 -6.80903931 -6.11589213 -6.11589213 -6.80903931 -5.71042702 -6.80903931
 -6.11589213 -5.71042702 -6.11589213 -5.19960139 -6.11589213 -6.80903931
 -6.80903931 -6.80903931 -6.11589213 -6.11589213 -6.80903931 -6.80903931
 -6.11589213 -6.11589213 -6.80903931 -6.809039

- error rate 확인!!

In [165]:
reload(bayes)

<module 'bayes' from 'bayes.pyc'>

In [166]:
bayes.spamTest()

classification error ['benoit', 'mandelbrot', '1924', '2010', 'benoit', 'mandelbrot', '1924', '2010', 'wilmott', 'team', 'benoit', 'mandelbrot', 'the', 'mathematician', 'the', 'father', 'fractal', 'mathematics', 'and', 'advocate', 'more', 'sophisticated', 'modelling', 'quantitative', 'finance', 'died', '14th', 'october', '2010', 'aged', 'wilmott', 'magazine', 'has', 'often', 'featured', 'mandelbrot', 'his', 'ideas', 'and', 'the', 'work', 'others', 'inspired', 'his', 'fundamental', 'insights', 'you', 'must', 'logged', 'view', 'these', 'articles', 'from', 'past', 'issues', 'wilmott', 'magazine']
pSpam:0.5
the error rate is:  0.1


- 일반적인 기계학습으로 시스템을 만들 때 testSet을 만들어 test를 하지!! 당연한 소리!! 
- 일반적인 기계학습은 그렇지만, 신경망 같은 경우는 훈련된 것에 누적해서 훈련을 한다!!! 1번 훈련 한 것을 Epoch 라고 한다!!! 

- p0v, p1v, pSpam 이 3가지 확률과 해당 단어 모음이 있으면 spam인지 ham인지 확률을 구할 수 있지!!! 

# 4.7 Example: using naïve Bayes to reveal local attitudes from personal ads

Example: using naïve Bayes to find locally used words
1. Collect: 
    - Collect from RSS feeds. We’ll need to build an interface to the RSS feeds.
2. Prepare: 
    - Parse text into token vectors.
3. Analyze: 
    - Inspect the tokens to make sure parsing was done correctly.
4. Train: 
    - Use trainNB0() that we created earlier.
5. Test: 
    - We’ll look at the error rate to make sure this is actually working. We can make modifications to the tokenizer to improve the error rate and results.
6. Use: 
    - We’ll build a complete program to wrap everything together. It will display the most common words given in two RSS feeds.

## 4.7.1 Collect: importing RSS feeds

In [169]:
import feedparser

In [175]:
ny=feedparser.parse('https://newyork.craigslist.org/stp/index.rss')

In [177]:
ny['entries']

[{'dc_source': u'https://newyork.craigslist.org/brk/stp/d/cool-friend-maybe-more-m4w/6311894226.html',
  'dc_type': u'text',
  'id': u'https://newyork.craigslist.org/brk/stp/d/cool-friend-maybe-more-m4w/6311894226.html',
  'language': u'en-us',
  'link': u'https://newyork.craigslist.org/brk/stp/d/cool-friend-maybe-more-m4w/6311894226.html',
  'links': [{'href': u'https://newyork.craigslist.org/brk/stp/d/cool-friend-maybe-more-m4w/6311894226.html',
    'rel': u'alternate',
    'type': u'text/html'}],
  'published': u'2017-09-19T02:56:54-04:00',
  'published_parsed': time.struct_time(tm_year=2017, tm_mon=9, tm_mday=19, tm_hour=6, tm_min=56, tm_sec=54, tm_wday=1, tm_yday=262, tm_isdst=0),
  'rights': u'copyright 2017 craiglist',
  'rights_detail': {'base': u'https://newyork.craigslist.org/search/stp?format=rss',
   'language': None,
   'type': u'text/plain',
   'value': u'copyright 2017 craiglist'},
  'summary': u"average guy seeking a cool female friend that I can chat with laugh bug out

In [178]:
len(ny['entries'])

25

In [179]:
ny['entries'][1]['summary']

u"Looking for Commitment - Let's Get Married!!! \nDon't Spend Another Day as a Single Woman! \nLet's see what we can do about that! \nI'm NOT looking for a one night'er or a fling, but rather a meaningful relationship. \nBased on Openness, Trust and Honest ..."

### Listing 4.6 RSS feed classifier and frequent word removal functions

In [ ]:
def calcMostFreq(vocabList,fullText):
    import operator
    freqDict = {}
    for token in vocabList:
        freqDict[token]=fullText.count(token)
    sortedFreq = sorted(freqDict.iteritems(), key=operator.itemgetter(1), \
                        reverse=True) # 가장 빈번하게 나온 순으로 정렬 
    return sortedFreq[:30]            # 그중 30개만!! num으로 parameter 받는게 좋음!! 

def localWords(feed1,feed0): # newyork feed / san francisco feed
    import feedparser
    docList=[]; classList = []; fullText =[]
    minLen = min(len(feed1['entries']),len(feed0['entries']))
    for i in range(minLen):
        wordList = textParse(feed1['entries'][i]['summary'])
        docList.append(wordList)
        fullText.extend(wordList)
        classList.append(1) #NY is class 1
        wordList = textParse(feed0['entries'][i]['summary'])
        docList.append(wordList)
        fullText.extend(wordList)
        classList.append(0)
    vocabList = createVocabList(docList) #create vocabulary
    
    #remove top 30 words
    top30Words = calcMostFreq(vocabList,fullText) 
    for pairW in top30Words:
        if pairW[0] in vocabList: vocabList.remove(pairW[0])
            
            
    #create train & test set
    trainingSet = range(2*minLen); testSet=[] 
    for i in range(20):
        randIndex = int(random.uniform(0,len(trainingSet)))
        testSet.append(trainingSet[randIndex])
        del(trainingSet[randIndex])
        
    trainMat=[]; trainClasses = []
    for docIndex in trainingSet:#train the classifier (get probs) trainNB0
        trainMat.append(bagOfWords2VecMN(vocabList, docList[docIndex]))
        trainClasses.append(classList[docIndex])
    p0V,p1V,pSpam = trainNB0(array(trainMat),array(trainClasses))
    # pSpam, 여기서는 newyork RSS일 확률이지!! 
    
    errorCount = 0
    for docIndex in testSet:        #classify the remaining items
        wordVector = bagOfWords2VecMN(vocabList, docList[docIndex])
        if classifyNB(array(wordVector),p0V,p1V,pSpam) != \
        classList[docIndex]:
            errorCount += 1

    print 'the error rate is: ',float(errorCount)/len(testSet)
    return vocabList,p0V,p1V

# 기존 코드와 동일하고 특이한 것은 TOP빈도 30개는 제거 했다는 것!! 

##### 실습

In [188]:
reload(bayes)

<module 'bayes' from 'bayes.py'>

In [189]:
ny=feedparser.parse('https://newyork.craigslist.org/stp/index.rss')
sf=feedparser.parse('https://sfbay.craigslist.org/stp/index.rss')


In [194]:
vocabList, pSF, pNY = bayes.localWords(ny,sf)

the error rate is:  0.35


In [191]:
len(ny['entries']), len(sf['entries'])

(25, 25)

- stop words는 빈도수가 많은 것으로 할 수도 있고, 미리 정해놓고 제할 수 도 있다!! 

## 4.7.2 Analyze: displaying locally used words

In [ ]:
def getTopWords(ny,sf):
    import operator
    vocabList,p0V,p1V=localWords(ny,sf)
    topNY=[]; topSF=[]
    for i in range(len(p0V)):
        if p0V[i] > -6.0 : topSF.append((vocabList[i],p0V[i]))
        if p1V[i] > -6.0 : topNY.append((vocabList[i],p1V[i]))
    sortedSF = sorted(topSF, key=lambda pair: pair[1], reverse=True)
    print "SF**SF**SF**SF**SF**SF**SF**SF**SF**SF**SF**SF**SF**SF**SF**SF**"
    for item in sortedSF:
        print item[0]
    sortedNY = sorted(topNY, key=lambda pair: pair[1], reverse=True)
    print "NY**NY**NY**NY**NY**NY**NY**NY**NY**NY**NY**NY**NY**NY**NY**NY**"
    for item in sortedNY:
        print item[0]

In [196]:
reload(bayes)

<module 'bayes' from 'bayes.py'>

In [202]:
bayes.getTopWords(ny,sf,5)

the error rate is:  0.25
SF**SF**SF**SF**SF**SF**SF**SF**SF**SF**SF**SF**SF**SF**SF**SF**
this
all
got
into
friend
NY**NY**NY**NY**NY**NY**NY**NY**NY**NY**NY**NY**NY**NY**NY**NY**
off
from
but
send
jerk
